### PR TITLE
Secretprovider interfaces and some default implementations

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation

This pr defines the SecretsProviderConfigurator that configures the function instances/containers and plugs in the right SecretsProvider for the instances to use.

### Modifications

Describe the modifications you've done.

### Result

After your change, what will change.

### CODE REVIEW

**Comment by srkukarni** [comment]
@cckellogg

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            Type type = new TypeToken<Map<String, Object>>() {
+            }.getType();
+            Map<String, Object> secretsMap = new Gson().fromJson(functionDetails.getSecretsMap(), type);
+            for (Map.Entry<String, Object> entry : secretsMap.entrySet()) {
+                final V1EnvVar secretEnv = new V1EnvVar();
+                Map<String, String> kv = (Map<String, String>) entry.getValue();
+                secretEnv.name(entry.getKey())
+                        .valueFrom(new V1EnvVarSource()
+                                .secretKeyRef(new V1SecretKeySelector()
+                                        .name(kv.entrySet().iterator().next().getKey())
```
Comment (cckellogg) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: should this look for specific keys in the Map? There should be some error checking? What if the first key is unexpected or unknown value? The provider should provide some validation of the secret object and not process it if it's not configured correctly. Can there be some unit tests with this too?

Comment (srkukarni) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: Thats a good question. For the default implementation, should this be just a simple (name, key) pair? And later, for more advanced usages, users will typically have their own implementation?

Comment (cckellogg) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: I think there still needs to be some validation for the simple case. This makes it easier to test and debug also and can provide information to the user who misconfigured the secret. Otherwise it's a black box and hard to debug.

Comment (srkukarni) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: I have added unittests as well as some sort of config checks.

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public Map<String, String> getSecretsProviderConfig(Function.FunctionDetails functionDetails) {
+        return null;
+    }
+
+    @Override
+    public void configureKubernetesRuntimeSecretsProvider(V1Container container, Function.FunctionDetails functionDetails) {
+        if (!StringUtils.isEmpty(functionDetails.getSecretsMap())) {
```
Comment (merlimat) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: Can you summarize in a comment what the implementation is doing here?

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        throw new RuntimeException("KubernetesSecretsProviderConfigurator should only be setup for Kubernetes Runtime");
+    }
+
+    @Override
+    public Type getSecretObjectType() {
+        return new TypeToken<Map<String, String>>() {}.getType();
+    }
+
+    @Override
+    public void validateSecretMap(Map<String, Object> secretMap) {
```
Comment (cckellogg) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: The secret map needs to be validated as well.
The two keys that are valid at this point are
"id" = the secret name in kubernetes
"key" = the key in that secret.
We need should have an instanceof check to confirm we are getting the correct object before the cast

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public void configureKubernetesRuntimeSecretsProvider(V1Container container, Function.FunctionDetails functionDetails) {
+        if (!StringUtils.isEmpty(functionDetails.getSecretsMap())) {
+            Type type = new TypeToken<Map<String, Object>>() {
+            }.getType();
+            Map<String, Object> secretsMap = new Gson().fromJson(functionDetails.getSecretsMap(), type);
+            for (Map.Entry<String, Object> entry : secretsMap.entrySet()) {
+                final V1EnvVar secretEnv = new V1EnvVar();
+                Map<String, String> kv = (Map<String, String>) entry.getValue();
+                secretEnv.name(entry.getKey())
```
Comment (cckellogg) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: final String secretName = kv.get("id")
final String serectKey = kv.get("key")

.name(secretName)
.key(secretKey)

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public Map<String, String> getSecretsProviderConfig(Function.FunctionDetails functionDetails) {
+        return null;
+    }
+
+    // Kubernetes secrets can be exposed as volume mounts or as environment variables in the pods. We are currently using the
+    // environment variables way. Essentially the secretName/secretPath is attached as secretRef to the environment variables
+    // of a pod and kubernetes magically makes the secret pointed to by this combination available as a env variable.
+    @Override
+    public void configureKubernetesRuntimeSecretsProvider(V1Container container, Function.FunctionDetails functionDetails) {
```
[the remark this answers is not in the document]
Comment (srkukarni) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: Not sure about that. The provider doesn't really know how the podspec is being structured by the runtime. For instance in the current implementation, there are two containers being created and this provider will not know about those details. All it needs to know is the container where the secrets need to be attached

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java (hunk 3 above)
Comment (srkukarni) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: done
